File: tests/st/ops/ascend/test_tensor_print/tensor_print_utils.py
# Copyright 2020 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
import numpy as np

import mindspore
import mindspore.context as context
import mindspore.nn as nn
from mindspore import Tensor
from mindspore.ops import operations as P

context.set_context(mode=context.GRAPH_MODE, device_target="Ascend")

class TensorPrint(nn.Cell):
    def __init__(self):
        super().__init__()
        self.print = P.Print()

    def construct(self, *inputs):
        self.print(*inputs)
        return inputs[0]

def get_tensor(is_scalar, input_type):
    if is_scalar == 'scalar':
        if input_type == mindspore.bool_:
            return Tensor(True, dtype=input_type)
        if input_type in [mindspore.uint8, mindspore.uint16, mindspore.uint32, mindspore.uint64]:
            return Tensor(1, dtype=input_type)
        if input_type in [mindspore.int8, mindspore.int16, mindspore.int32, mindspore.int64]:
            return Tensor(-1, dtype=input_type)
        if input_type in [mindspore.float16, mindspore.float32, mindspore.float64]:
            return Tensor(0.01, dtype=input_type)
    else:
        if input_type == mindspore.bool_:
            return Tensor(np.array([[True, False], [False, True]]), dtype=input_type)
        if input_type in [mindspore.uint8, mindspore.uint16, mindspore.uint32, mindspore.uint64]:
            return Tensor(np.array([[1, 2, 3], [4, 5, 6]]), dtype=input_type)
        if input_type in [mindspore.int8, mindspore.int16, mindspore.int32, mindspore.int64]:
            return Tensor(np.array([[-1, 2, -3], [-4, 5, -6]]), dtype=input_type)
        if input_type in [mindspore.float16, mindspore.float32, mindspore.float64]:
            return Tensor(np.array([[1.0, -2.0, 3.0], [4.0, -5.0, 6.0]]), dtype=input_type)
    return Tensor(False, np.bool)

if __name__ == "__main__":
    net = TensorPrint()
    net(get_tensor('scalar', mindspore.bool_), get_tensor('scalar', mindspore.uint8),
        get_tensor('scalar', mindspore.int8), get_tensor('scalar', mindspore.uint16),
        get_tensor('scalar', mindspore.int16), get_tensor('scalar', mindspore.uint32),
        get_tensor('scalar', mindspore.int32), get_tensor('scalar', mindspore.uint64),
        get_tensor('scalar', mindspore.int64), get_tensor('scalar', mindspore.float16),
        get_tensor('scalar', mindspore.float32), get_tensor('scalar', mindspore.float64),
        get_tensor('array', mindspore.bool_), get_tensor('array', mindspore.uint8),
        get_tensor('array', mindspore.int8), get_tensor('array', mindspore.uint16),
        get_tensor('array', mindspore.int16), get_tensor('array', mindspore.uint32),
        get_tensor('array', mindspore.int32), get_tensor('array', mindspore.uint64),
        get_tensor('array', mindspore.int64), get_tensor('array', mindspore.float16),
        get_tensor('array', mindspore.float32), get_tensor('array', mindspore.float64))

File: mindspore/ccsrc/plugin/device/ascend/hal/device/tensorprint_utils.cc
/**
 * Copyright 2020-2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "plugin/device/ascend/hal/device/tensorprint_utils.h"
#include <fstream>
#include <iostream>
#include <memory>
#include <string>
#include "pybind11/pybind11.h"
#include "ir/tensor.h"
#include "utils/file_utils.h"
#include "utils/log_adapter.h"
#include "proto/print.pb.h"
#include "utils/ms_context.h"

namespace py = pybind11;

namespace mindspore::device::ascend {
namespace {
void OutputReceiveData2StdOut(const ScopeAclTdtDataset &dataset) {
  // Acquire Python GIL
  py::gil_scoped_acquire gil_acquire;

  for (auto data_elem : dataset.GetDataItems()) {
    if (std::holds_alternative<std::string>(data_elem)) {
      std::cout << std::get<std::string>(data_elem) << std::endl;
    } else {
      auto tensor_ptr = std::get<mindspore::tensor::TensorPtr>(data_elem);
      std::cout << tensor_ptr->ToStringNoLimit() << std::endl;
    }
  }
}
}  // namespace

TensorPrintUtils &TensorPrintUtils::GetInstance() {
  static TensorPrintUtils instance;
  return instance;
}

TensorPrintUtils::TensorPrintUtils() {
  auto ms_context = MsContext::GetInstance();
  MS_EXCEPTION_IF_NULL(ms_context);
  print_file_path_ = ms_context->get_param<std::string>(MS_CTX_PRINT_FILE_PATH);
  if (print_file_path_.empty()) {
    return;
  }
  ChangeFileMode(print_file_path_, S_IWUSR);
  pb_file_stream_ =
    std::make_shared<std::fstream>(print_file_path_, std::ios::out | std::ios::trunc | std::ios::binary);
}

TensorPrintUtils::~TensorPrintUtils() {
  if (pb_file_stream_ != nullptr) {
    pb_file_stream_->close();
    ChangeFileMode(print_file_path_, S_IRUSR);
  }
}

void TensorPrintUtils::PrintReceiveData(const ScopeAclTdtDataset &dataset) {
  if (pb_file_stream_ == nullptr) {
    OutputReceiveData2StdOut(dataset);
  } else {
    // output data to file in protobuf binary format
    OutputReceiveData2PbFile(dataset);
  }
}

void TensorPrintUtils::OutputReceiveData2PbFile(const ScopeAclTdtDataset &dataset) {
  prntpb::Print print;

  for (auto data_elem : dataset.GetDataItems()) {
    prntpb::Print_Value *value = print.add_value();

    if (std::holds_alternative<std::string>(data_elem)) {
      value->set_desc(std::get<std::string>(data_elem));
      continue;
    }

    auto tensor_ptr = std::get<mindspore::tensor::TensorPtr>(data_elem);
    prntpb::TensorProto *tensor = value->mutable_tensor();
    for (const auto &dim : tensor_ptr->shape()) {
      tensor->add_dims(static_cast< ::google::protobuf::int64>(dim));
    }

    tensor->set_tensor_type(tensor_ptr->Dtype()->ToString());
    tensor->set_tensor_content(tensor_ptr->data_c(), tensor_ptr->data().nbytes());
  }

  if (!print.SerializeToOstream(pb_file_stream_.get())) {
    MS_LOG(ERROR) << "Save print file:" << print_file_path_ << " fail.";
  }
  print.Clear();
}
}  // namespace mindspore::device::ascend
